usr/share/lib/img_proof/tests/SLES/test_sles_hardened.py

```diff
@@ -0,0 +1,59 @@
+import pytest
+
+
+# Workaround for https://github.com/OpenSCAP/openscap/issues/1796
+def setup_swap(host, swap_file):
+    mem_total = int(host.run(
+        "cat /proc/meminfo | awk '$0 ~ /MemTotal:/ { print $2 }'"
+    ).stdout.strip())
+    if mem_total < 16e6:
+        for command in [
+            "sudo fallocate -l 4G {}".format(swap_file),
+            "sudo chmod 600 {}".format(swap_file),
+            "sudo mkswap {}".format(swap_file),
+            "sudo swapon {}".format(swap_file),
+        ]:
+            result = host.run(command)
+            if result.rc != 0:
+                raise RuntimeError(result.stderr.strip())
+
+
+def test_sles_hardened(host, get_release_value, is_sles_sap):
+    variant = get_release_value('VARIANT_ID') or ''
+
+    if "hardened" not in variant:
+        pytest.skip('Skipping non-hardened image')
+
+    swap_file = "/tmp/swap"
+    try:
+        setup_swap(host, swap_file)
+        xml_path = "pub/projects/security/oval/suse.linux.enterprise.15.xml"
+        # Downloaded file should have slashes replaced by hyphens
+        xml_file = xml_path.replace("/", "-")
+        oscap_dir = "/tmp/oscap"
+        oscap_file = "/usr/share/xml/scap/ssg/content/ssg-sle15-ds.xml"
+        oscap_profile = "pcs-hardening-sap" if is_sles_sap() else "pcs-hardening"
+
+        host.run("mkdir {}".format(oscap_dir))
+        host.run(
+            "curl -o- https://ftp.suse.com/{path}.gz | gunzip -c > {dir}/{file}".format(
+                path=xml_path,
+                file=xml_file,
+                dir=oscap_dir,
+            )
+        )
+
+        result = host.run(
+            "sudo oscap xccdf eval --local-files {dir} --profile {profile} {file}".format(
+                dir=oscap_dir,
+                file=oscap_file,
+                profile=oscap_profile,
+            )
+        )
+    except RuntimeError as exc:
+        pytest.fail("Test failed with %s" % exc)
+    finally:
+        host.run("sudo swapoff {}".format(swap_file))
```

Comment:
If these move to a separate function they could be re-used here and in the setup_swap function.

Reply:
Much better to use a context manager for this.

Reply:
We keep changing multiple things at once and ending up in situations where the test does not run as intended. Contextmanager is fine but it's not bullet proof. It also can cause trouble inside pytest.

With latest changes there's multiple paths that lead to "deadlock". The interaction between raising exception, context manager and calling pytest.skip/fail is leading pytest to lock up. This would be very bad in automated testing. I don't know the exact mechanism at play but it might be related to the context manager not yielding properly. Anyway, sticking to a simpler solution will limit chances we hit this type of scenario.

As far as I can tell there's two main things we are trying to achieve.

- Always cleanup no matter pass/fail/exception
- Skip test if the instance doesn't have enough memory and fails to setup sufficient swap space

Suggest a couple changes.

- Remove the try/except around the context manager. This is not needed. If there's an unhandled exception, so long as cleanup happens the test should fail. Pytest already handles exceptions as failures.
- Return a result code from context manager rather than raising exception
- Handle result code from context manager and skip outside of context

Example context manager:

```
    rc = 0
    try:
        if mem_total < min_memory:
            for command in [
                "sudo fallocate -l 4G {}".format(swap_file),
                "sudo chmod 600 {}".format(swap_file),
                "sudo mkswap {}".format(swap_file),
                "sudo swapon {}".format(swap_file),
            ]:
                result = host.run(command)
                if result.rc != 0:
                    rc = result.rc
                    break
        yield rc
    finally:
        ...
```

Example code:

```
    skipped = False
    with setup_swap(host) as rc:
        if rc != 0:
            skipped = True
        else:
            ... # do testing

    if skipped:
        pytest.skip(
            "Failed to setup swap, not enough memory."
        )

    assert result.rc == 0
```

Reply:
We want to avoid any interaction where pytest.skip/fail is called inside a try/except block or inside a context manager. And ditto for exceptions in context manager body prior to yield.

Reply:
Fixed

Reply:
Ran the latest iteration against an SAP hardened image and the tests fails on exception:

```
Exception ignored in: <generator object setup_swap at 0x7fcadc25d7e0>
Traceback (most recent call last):
  File "/home/smarlow/projects/img_proof/usr/share/lib/img_proof/tests/SLES/test_sles_hardened.py", line 30, in setup_swap
  File "/home/smarlow/projects/venvs/img_proof/lib64/python3.11/site-packages/testinfra/host.py", line 88, in run
  File "/home/smarlow/projects/venvs/img_proof/lib64/python3.11/site-packages/testinfra/backend/paramiko.py", line 142, in run
  File "/home/smarlow/projects/venvs/img_proof/lib64/python3.11/site-packages/testinfra/backend/paramiko.py", line 124, in _exec_command
  File "/home/smarlow/projects/venvs/img_proof/lib64/python3.11/site-packages/testinfra/utils/__init__.py", line 29, in __get__
  File "/home/smarlow/projects/venvs/img_proof/lib64/python3.11/site-packages/testinfra/backend/paramiko.py", line 100, in client
NameError: name 'open' is not defined
```

This seems to be an issue between pytest/paramiko and the contextmanager. Given pytest does a bunch of monkey patching this is not surprising.

Reply:
Hopefully fixed now.

Reply:
Yep, thanks